### PR TITLE
Map user memory passed to accept() in kernel build

### PR DESCRIPTION
**Pull request by gpoulios**
## Summary

At least on RISC-V, TCP server sockets were pretty much unusable on high load environments and kernel build.

User addresses passed to `accept()` would be accessed when the wrong MMU mappings were active. A crash would manifest when attempting to `accept()` for instance on a TCP server socket under significant load. The accept event handler would be called by the HP worker upon client connection. At this point, `accept_tcpsender()` would attempt to write to `addr` resulting in a page fault. Reproducibility would depend on the current system load (num tasks or CPU stress) but in loaded environments, it would crash almost 100% of the times.

I have not reproduced the issue on other boards/archs, but I think it should be applicable:
 - `accept()` gets called by task A
 - context is switched to task B
 - client attempts to connect (interrupt)
 - HP worker processes the new connection and invokes the accept event callback (eventually `accept_tcpsender()`)
 - `accept_tcpsender()` attempts to write to user memory
   - -> no guarantee that user memory is mapped at this point

## Impact

 - **Is new feature added**? NO
 - **Is existing feature changed**? YES, with `CONFIG_BUILD_KERNEL=y`, the `accept()` function now handles user-provided addresses more robustly.
 - **User Impact**: NO, the change should be transparent to users and improve stability. Effective changes only with `CONFIG_BUILD_KERNEL=y`.
 - **Build Impact**: NO
 - **Hardware Impact**: RISC-V affected by issue. Other architectures potentially affected but untested. Fix applicable to all architectures.
 - **Documentation Impact**: NO
 - **Security Impact**: Not really: Mostly a crash and even though it has potential for corruption of others' memory with semi-controlled data and controlled address, it's insignificant in the grand scheme of things as this is also the case for many other NuttX syscalls. Fix enhances stability during applicable circumstances, and ever so slightly the security by checking that pointers actually belong to user (and not the kernel, for instance).
 - **Compatibility Impact**: NO

It should be noted that Linux does this the other way around: it operates on kernel stack allocated data and once done, it copies them to user. This can also be a viable alternative, albeit with one extra copy and a little extra memory.

### Tested only on RISC-V (Microchip Polarfire)

Prior to patch: trying with `CONFIG_BUILD_KERNEL=y` and `CONFIG_NETUTILS_NETCAT=y` would cause a store fault once a client attempted to connect. The point of the crash was consistently `tcp_accept.c / accept_tcpsender()` at the first write to the `inaddr` structure ([line 102](https://github.com/apache/nuttx/blob/5d68ab635c065d97ac0cbc08372be7c37fb59d02/net/tcp/tcp_accept.c#L102)):

```
knsh> netcat -l 9999
log: net: listening on :9999
riscv_exception: EXCEPTION: Store/AMO page fault. MCAUSE: 000000000000000f, EPC: 00000000a002f4d2, MTVAL: 00000000c000af90
[...]
dump_assert_info: Assertion failed panic: at file: :0 task: hpwork process: Kernel 0xa0022594
```
...with`0xc000af90` being the 2nd argument passed to `accept()`.

After the patch: `accept()` works normally after multiple connections.

### CODE REVIEW

**Comment by nuttxpr** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



This PR description mostly meets the NuttX requirements, but could be improved in a few areas.

**Strengths:**

* **Clear Summary:** The summary clearly explains the problem, the affected area, and the solution's approach.
* **Impact Description:**  The impact section adequately addresses most points, especially user impact and potential alternatives.
* **Testing Information:** Provides specific details on the testing environment and observed behavior before and after the change, including relevant log snippets and code references.

**Areas for Improvement:**

* **Summary:**  While the summary mentions RISC-V, it should explicitly state if this fix *only* applies to RISC-V or if it's a general TCP fix that was just tested on RISC-V.  Be explicit about the scope of the change.  Add the related NuttX issue number if one exists.
* **Impact:**
    * **Build Impact:**  Even if there is no build impact, explicitly state "NO".  Same for all other "Impact" items.
    * **Hardware Impact:** While it mentions RISC-V, clarify if *only* RISC-V is affected. If other architectures are potentially affected but untested, mention that explicitly.  Be precise about the affected hardware/architectures.
    * **Documentation Impact:**  Even if no documentation update is required, state "NO". If the change warrants a documentation update (even a small one), do it and mention it here as "YES" with a brief description.
    * **Security Impact:** While it doesn't appear to be a direct security issue, explicitly state "NO" or if there are subtle security implications due to the previous faulty behavior, briefly describe them.
    * **Compatibility Impact:** Explicitly state "NO" if no compatibility issues are expected.

* **Testing:**
    * **Build Host(s):** Provide more detail. E.g., "Linux (Ubuntu 20.04), x86_64, GCC 9.4.0".
    * **Target(s):** Be more specific. E.g., "RISC-V (QEMU virt), sim:nsh".  If only tested on RISC-V, say so explicitly.  If applicable to other architectures but untested, state that clearly.
    * **Conciseness of Logs:** While the provided logs demonstrate the issue, trim them to the most relevant parts for brevity. Focus on the lines immediately before and after the crash, and any key messages related to the fix.


**Example of improved sections:**

**Summary:**

Fix for TCP server socket instability under high load on RISC-V.  Addresses a potential crash in `accept()` due to incorrect memory access when handling client connections. The fix maps user-provided address pointers to kernel virtual memory before use and validates that they point to user space.  Relates to NuttX Issue #[Insert Issue Number if applicable].

**Impact:**

* Is new feature added? NO
* Is existing feature changed? YES, the `accept()` function now handles user-provided addresses more robustly.
* Impact on user: NO. The change should be transparent to users and improve stability.
* Impact on build: NO
* Impact on hardware: YES, this fix specifically targets RISC-V. Other architectures might have similar vulnerabilities but have not been tested in this PR.
* Impact on documentation: NO
* Impact on security: NO (While the previous behavior could lead to crashes, it did not directly expose a security vulnerability).
* Impact on compatibility: NO


**Testing:**

* Build Host(s): Linux (Ubuntu 22.04), x86_64, GCC 11.2.0
* Target(s): RISC-V (QEMU virt), sim:nsh

Testing logs before change (truncated):

```
knsh> netcat -l 9999
log: net: listening on :9999
riscv_exception: EXCEPTION: Store/AMO page fault. MCAUSE: [...] EPC: [...] MTVAL: 00000000c000af90  <-- User address
```

Testing logs after change (concise summary):

`netcat -l 9999` now accepts multiple client connections without crashing. Tested with [number] concurrent connections.



By addressing these points, the PR will be clearer, more complete, and easier for reviewers to assess.

File: fs/socket/accept.c [hunk 1]
```diff
@@ -137,6 +191,36 @@ int accept4(int sockfd, FAR struct sockaddr *addr, FAR socklen_t *addrlen,
...
     }
 
+  /* Map user addresses to kernel virtual memory and check they actually
+   * belong to the user
+   *
+   * NOTE: also check if writeable?
+   * NOTE: possible TOCTOU with contents of addrlen?
+   */
+#ifdef CONFIG_MM_KMAP
+
```
Comment (pussuw) [inline fs/socket/accept.c:201]: Remove extra linefeeds after #ifdef #else and before #endif directive.

File: fs/socket/accept.c [hunk 2]
```diff
@@ -137,6 +191,36 @@ int accept4(int sockfd, FAR struct sockaddr *addr, FAR socklen_t *addrlen,
       goto errout;
     }
 
+  /* Map user addresses to kernel virtual memory and check they actually
+   * belong to the user
+   *
+   * NOTE: also check if writeable?
+   * NOTE: possible TOCTOU with contents of addrlen?
+   */
```
Comment (pussuw) [inline fs/socket/accept.c:199]: Add linefeed after comment line

File: fs/socket/accept.c [hunk 3]
```diff
@@ -38,10 +38,62 @@
...
+ *  Returns 0 (OK) on success, or:
+ *    -ENOMEM
+ *      If there is not enough free memory to create the mapping.
+ *    -EFAULT
+ *      If `uaddr` points to memory not belonging to the user address space.
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_MM_KMAP
+static int map_user_mem(FAR void *uaddr, size_t size, void **p_kaddr)
```
Comment (pussuw) [inline fs/socket/accept.c:74]: If the purpose of this function is to test if user has provided a valid user address (and not trying to give a kernel address for attacks), then it is unnecessary. The user given pointer checks should be more comprehensive i.e. on the system call layer for ALL SYSTEM CALLS (which we don't have yet). Regardless, this is for sure not the correct position to do such checks.

Comment (gpoulios) [reply fs/socket/accept.c:74]: The purpose of the function is to map user memory to kernel addressable memory.

In order to do that, `kmm_map_user()` also checks whether it is indeed the given task's memory it's trying to map, and if not, it returns non-NULL result. I reckon this is a desired error handling to do in this situation. 

AFAIK there is no legitimate use case for passing any other kind of memory (other task's, or kernel's address space) to this function - with `CONFIG_BUILD_KERNEL=y`, which `CONFIG_MM_KMAP` depends on.

Regarding the syscall layer, the function itself (along with others) might indeed be placed in a more accessible unit, and be exported to all syscall handlers. The actual checking using such functions though, not so easy; perhaps the first level pointers through some metadata (e.g. argN is a pointer to size `<sz>`), but nested types (structs with pointers etc) would be more tricky. Btw, Linux does manual checking in a very similar fashion to this PR (see `accept()`'s [invocation](https://github.com/torvalds/linux/blob/eabcdba3ad4098460a376538df2ae36500223c1e/net/socket.c#L1951) and [body](https://github.com/torvalds/linux/blob/eabcdba3ad4098460a376538df2ae36500223c1e/net/socket.c#L274-L299) of `move_addr_to_user()`). Not saying we can't do better, just that it is tricky.

Bottomline:
 - if you want to see the `EFAULT` checking removed please let me know what to do with this case (nothing else makes sense to me right now but I'll do anything to see it merged).
 - if you want to see `map_user_mem()` in some place else, please name it and I'll put it there along with any other changes you request.

Comment (pussuw) [reply fs/socket/accept.c:74]: Thanks for the thorough explanation. I'm fine with it as long as there is a clear motivation for it.

**Review by pussuw — APPROVED** [review]

File: fs/socket/accept.c [hunk 4]
```diff
@@ -38,10 +38,62 @@
...
+
+  if (uaddr)
+    {
+      *p_kaddr = kmm_map_user(this_task(), uaddr, size);
+
+      if (*p_kaddr == NULL)
+        {
+          return -ENOMEM;
+        }
+
```
Comment (acassis) [inline fs/socket/accept.c:86]: Please remove this empty line

**Review by acassis — APPROVED** [review]

**Review by acassis — APPROVED** [review]

**Review by pussuw — APPROVED** [review]

**Pull request merged**